syntax = "proto3";

package rpc;

service FileSystem {
  rpc CreateFile(Filename) returns (Response) {}
  rpc DeleteFile(Filename) returns (Response) {}
  rpc ListFiles(Empty) returns (stream FileList) {}
  rpc Upload(stream UploadRequest) returns (Response) {}
  rpc Download(Filename) returns (stream Chunk) {}
  rpc Lock(Filename) returns (Response) {}
  rpc Unlock(Filename) returns (Response) {}
}

message Empty {}

message Chunk {
  bytes buffer = 1;
}

message UploadRequest {
  oneof test_oneof{
    string filename = 1;
    bytes buffer = 2;
  }
}

enum StatusCode {
  OK = 0;
  Failed = 1;
}

message Response {
  StatusCode code = 1;
  string message = 2;
}

message Filename {
  string filename = 1;
}

message FileList {
  string filename = 1;
  float size = 2;
  float createdtime = 3;
  float modifiedtime = 4;
}